blog/_posts/2025-07-15-network-perf-aks.md

```diff
@@ -0,0 +1,89 @@
+
+![image](/assets/images/network-perf-aks/buffer_throughput.png)
+
+We also observed reductions in both TCP retransmission rate and round-trip time (RTT) under a combined bandwidth load of 15 Gbps. The benefit was especially pronounced when traffic was split across three parallel TCP streams, each operating at 5 Gbps.
+
+![image](/assets/images/network-perf-aks/buffer_retransmits.png)
+
+![image](/assets/images/network-perf-aks/buffer_rtt.png)
+
+TCP retransmissions often occur when the sender or receiver buffer is too small to handle a surge of packets, resulting in packet drops. For example, if the NIC name is `enP28334s1` (typical name for an [SR-IOV network interface](https://learn.microsoft.com/en-us/windows-hardware/drivers/network/overview-of-single-root-i-o-virtualization--sr-iov-)), you can check how many packets were dropped due to ring buffer overflow with:
+
+```bash
+ethtool -S enP28334s1 | grep rx_out_of_buffer
+```
+
+If `rx_out_of_buffer` shows a non-zero value, it indicates that a ring buffer overflow has occurred. To increase the ring buffer size (e.g., to 2048 packets), use:
+
+```bash
+sudo ethtool -G enP28334s1 rx 2048
+```
+
+Again you can also deploy a daemonSet to enforce consistent ring buffer size across all nodes in AKS cluster following [this example](https://github.com/Azure/telescope/blob/main/modules/kustomize/ring_buffer/ring-buffer-config.yaml).
+
+It’s important to note that increasing the NIC ring buffer size has memory usage implications. Allocating larger buffers means the operating system reserves more memory specifically for handling network traffic, which reduces the amount of memory available for user-space applications. For example, doubling the receive buffer size from 1024 to 2048 packets per descriptor across multiple queues and interfaces can lead to a non-trivial (around 100 MB or more) increase in kernel memory usage — especially on systems with high network concurrency or multiple high-throughput NICs. While this tradeoff can significantly improve network performance, especially under high load, it should be balanced against the memory demands of the application workload running on the same VM.
+
+## Conclusion
+
+Achieving optimal network performance on AKS requires a combination of choosing the right VM SKU and series  and fine-tuning kernel-level parameters. By understanding the trade-offs and evaluating different options, AKS users can unlock meaningful improvements in network performance and application responsiveness. In the future, we plan to expand our benchmarks to include newer Linux 6.x kernels, which incorporate recent [networking enhancements](https://conferences.computer.org/sc-wpub/pdfs/SC-W2024-6oZmigAQfgJ1GhPL0yE3pS/555400a775/555400a775.pdf). We also plan to measure and analyze the performance implications of container networking solutions (such as Cilium), and to explore additional optimization strategies.
```

Comment:
Could we close with something that invites comments for things Cx would like to see in these benchmarks?

Reply:
Done